pkg/controller/upgradeconfig/upgradeconfig_controller.go

```diff
@@ -157,5 +157,18 @@ func (r *ReconcileUpgradeConfig) Reconcile(request reconcile.Request) (reconcile
 
 // TODO readyToUpgrade checks whether it's ready to upgrade based on the scheduling
 func isReadyToUpgrade(upgradeConfig *upgradev1alpha1.UpgradeConfig) bool {
-	return true
+	if !upgradeConfig.Spec.Proceed {
+		log.Info("upgrade cannot be proceed", "proceed", upgradeConfig.Spec.Proceed)
+		return false
+	}
+	upgradeTime, err := time.Parse(time.RFC3339, upgradeConfig.Spec.UpgradeAt)
+	if err != nil {
+		log.Error(err, "failed to parse spec.upgradeAt", upgradeConfig.Spec.UpgradeAt)
+		return false
+	}
+	now := time.Now()
+	if now.After(upgradeTime) && upgradeTime.Add(15*time.Minute).After(now) {
```

Comment:
This should be an or statement?

Reply:
I had to scratch my head a bit at the logic too, but it checks out (assuming it's the behaviour we want).
It's basically saying "If the current time is after the scheduled time, and is also 15 minutes after the upgrade time, don't upgrade". ie. if the operator is more than 15 minutes late getting around to the request, we're not going to upgrade.
It's the same thing I was getting at on Slack really. I feel a bit uneasy putting that policy in code though. But we do need to clearly define what the operator's behaviour should be - ie. how "late" is too late when commencing an upgrade?